In [1]:
import validaciones

# Ejercicios Python

## Zen of Python

Lee una vez mas el Zen of Python antes de iniciar con la tarea

In [2]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


## 1 - XOR

En la siguiente celda crea una funcion para la compuerta logica XOR 

<img src="assets/xor.png" width="400">


In [3]:
def xor(a,b):
    """
    Esta funcion recibe dos valores booleanos, ejemplo xor(1, 0) o True y False
    y retorna 0 o 1 como resultado de la operacion logica XOR
    """
    # tu codigo aqui
def xor(a,b):
    return bool(a) ^ bool(b)

In [4]:
xor(0 , 1) 

True

In [5]:
xor(0 , 0)

False

In [6]:
xor(True , False)

True

### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

In [7]:
validaciones.xor_function(xor)

Prueba 1
Prueba 1 [CORRECTA]
Prueba 2
Prueba 2 [CORRECTA]


## 2 - Sumando N series

Tenemos una secuencia cuyo $n^{th}$ es:
\begin{equation*}
\ T_n = n^2 - (n-1)^2
\end{equation*}

Hay que evaluar las series:
\begin{equation*}
\ S_n = T_1 + T_2 + ... + T_n
\end{equation*}


**Instrucciones**
En la siguiente celda en la funcion summingSeries debemos **retornar** en valor de $S_n$, tenemos como entrada n

**Ejemplo 1:**

Si la entrada es 2

Salida:
4

Explicacion:

$T_1 = 1$

$T_2 = 3$

$S_2 = T_1 + T_2$

$S_2 = 4 $



**Tip :**
Antes te implementar $S_n$ la solución, analicemos algebraicamente el problema para encontrar una solución equivalente optima.


In [32]:
def summatoriaSerie(n):
    """
    La funcion recibe como parametro n y retorna el valor se la serie
    """
    # tu codigo aqui
    y=n**2
    return y

summatoriaSerie(10)

100

#### 2.1 - Ciclos

Crea un ciclo for que imprima el resultado de la funcion **summingSeries** desde 1 hasta 10

In [33]:
n=11
for n in range (1,n):
    print(summatoriaSerie(n))
    

1
4
9
16
25
36
49
64
81
100


### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

##### Importante :
En este ejercicio se evalua que la funcion sea optima, el tiempo de ejecucion tiene que ser el menor posible, se va a evaluar el resultado correcto y el tiempo de ejecucion, si en alguna de las pruebas obtienes **[FALLO: TIMEOUT]** la funcion no es optima debes cambiarla hasta que todas las pruebas pasen

In [16]:
validaciones.summingSeries_function(summatoriaSerie)

Prueba 1
Prueba 1[CORRECTA]
Prueba 2
Prueba 2[CORRECTA]
Prueba 3
Prueba 3[CORRECTA]
Prueba 4
Prueba 4[CORRECTA]
Prueba 5
Prueba 5[CORRECTA]
Prueba 6
Prueba 6[CORRECTA]


## 3 - Factores Primos

Para cada $n$ queremos obtener el conteo maximo de numeros primos unicos en en rango inclusivo $[1, n]$,
y retornar el valor del conteo en una nueva linea

**Nota:** Recuerda que un numero primo solo es divisible por el mismo y que 1 no es un numero primo

##### Ejemplos:

###### Ejemplo 1
Entrada: 1

Salida esperada: 1

Explicacion: El numero maximo de factores primos unicos en el rango inclusivo $[1,1]$ es $0$, porque $1$ no es un numero primo.

###### Ejemplo 2
Entrada: 3

Salida esperada: 1

Explicacion: El numero maximo de factores primos unicos en el rango inclusivo $[1,3]$ es $1$, porque el numero $3$ tiene 1 factor primo unico (el mismo) 

###### Ejemplo 3
Entrada: 500

Salida esperada: 4

Explicacion: El numero maximo de factores primos unicos en el rango inclusivo $[1,500]$ es $4$, porque el producto de los primeros cuatro numeros primos unicos es $2 \times 3 \times 5 \times 7 = 210$ y no hay ningun numero primo unico que multiplicado por el resultado sea $	\leqslant 500$



#### Tip: 
Utiliza la funcion `range()` de python


In [13]:
for i in range (1,500,1):
    print(i)

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
277


In [11]:
def conteoPrimos(n):
    """
    Esta funcion recibe como parametro n y 
    retorna el conteo maximo de numeros primos unicos
    """
    # tu codigo aqui

### Pruebas

In [12]:
conteoPrimos(1)

0

In [13]:
conteoPrimos(500)

4

### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

##### Importante :
En este ejercicio se evalua que la funcion sea optima, el tiempo de ejecucion tiene que ser el menor posible, se va a evaluar el resultado correcto y el tiempo de ejecucion, si en alguna de las pruebas obtienes **[FALLO: TIMEOUT]** la funcion no es optima debes cambiarla hasta que todas las pruebas pasen

In [14]:
validaciones.primeCount_function(conteoPrimos)

Prueba 1
Prueba 1[CORRECTA]
Prueba 2
Prueba 2[CORRECTA]
Prueba 3
Prueba 3[CORRECTA]
Prueba 4
Prueba 4[CORRECTA]
Prueba 5
Prueba 5[CORRECTA]
Prueba 6
Prueba 6[CORRECTA]
Prueba 7
Prueba 7[CORRECTA]
Prueba 8
Prueba 8[CORRECTA]
Prueba 9
Prueba 9[CORRECTA]
Prueba 10
Prueba 10[CORRECTA]
Prueba 11
Prueba 11[CORRECTA]
Prueba 12
Prueba 12[CORRECTA]


## 4 - Metodo de Euler

El metodo de Euler se utiliza para aproximar la solucion particular de una ecuacion diferencial dado un valor inicial. Con esta informacion se sabe que la grafica de esa solucion pasa a traves del punto $(x_0,y_0)$ y tiene una pendiente $F(x_0,y_0)$ en ese punto, esto da un punto de partida para aproximar la solucion.

A partir del punto inicial, se sigue en la direccion indicada por la pendiente. Mediante un pequeño paso $h$, se mueve a lo largo de la recta tangente hasta llegar al punto $(x_1,y_1)$ donde:
\begin{equation*}
\ x_1 = x_0 + h       \\
\ y_1 = y_0 + hF(x_0,y_0)\\
\end{equation*}


<img src="assets/Euler.png" width="350">

Como se muestra en la figura se considera $(x_1,y_1)$ como un nuevo punto inicial.

Los valores de $x$ son:
\begin{equation*}
\ x_1 = x_0 + h       \\
\ x_2 = x_1 + h       \\
\    .                \\
\    .                \\
\    .                \\
\ x_n = x_{n-1} + h   \\
\end{equation*}


Los valores de $y$ son:
\begin{equation*}
\ y_1 = y_0 + hF(x_0,y_0)\\
\ y_1 = y_1 + hF(x_1,y_1)\\
\       . \\
\       . \\
\       . \\
\ y_n = y_{n-1} + hF(x_{n-1},y_{n-1})\\
\end{equation*}


#### Problema 4
El ejecicio consiste en implementar la funcion *euler_metod* esta funcion recibe como parametro el 
* punto inicial $x_0$
* punto inicial $y_0$
* Paso $h$ 
* numero de iteraciones $n$ 
* La equacion diferencial que queremos aproximar, esta funcion tambien recibe los parametros $x$ y $y$
* La funcion de la solucion exacta que tambien tambien recibe los parametros $x$ y $y$

La funcion debe retornar el valor de $y_n$ y el error
$error=|y - y_n|$ 

In [15]:
import math 

In [16]:
def euler_method(x,y,h,n, diff_eq, func):
    """
    Parametros x, y punto inicial, logitud del paso h , n numero de iteraciones,
    diff_eq la equacion diferencial y func la funcion de la solucion exacta
    Retorna y y el error
    """
    # tu codigo aqui
        
   
        

### Ejemplo

Queremos aproximar la solucion particular de la ecuacion diferencial
\begin{equation*}
\ \frac{\partial y}{\partial x} = -2y \\
\end{equation*}
Donde $y(0) = 4$ vamos a unsa un de $h=0.1$

Tambien vamos a calcular el error en este caso podemos verificar el error  con la solucion exacta:
\begin{equation*}
\ y = 4e^{-2x} \\
\end{equation*}

In [17]:
### Implementando las funciones del ejemplo
def y_prime(x, y):
    return -2*y

def y(x):
    return 4 * math.exp(-2*x)

In [18]:
"""
Llamamos a la funcion de euler
euler_method(0,4,0.1,10, y_prime,y)
Parametros :
x = 0
y = 4
d = 0.1
n = 10
diff_eq = y_prime
func = y
"""

euler_method(0,4,0.1,10, y_prime,y)

x:  0.1
y_1:  3.2
y:  3.2749230123119273
error:  0.0749230123119271
x:  0.2
y_2:  2.56
y:  2.6812801841425573
error:  0.12128018414255726
x:  0.3
y_3:  2.048
y:  2.195246544376106
error:  0.14724654437610596
x:  0.4
y_4:  1.6384
y:  1.7973158564688863
error:  0.15891585646888617
x:  0.5
y_5:  1.3107
y:  1.4715177646857693
error:  0.16081776468576936
x:  0.6
y_6:  1.0486
y:  1.2047768476488085
error:  0.15617684764880857
x:  0.7
y_7:  0.8389
y:  0.986387855766426
error:  0.14748785576642598
x:  0.8
y_8:  0.6711
y:  0.8075860719786215
error:  0.1364860719786215
x:  0.9
y_9:  0.5369
y:  0.6611955528863461
error:  0.12429555288634608
x:  1.0
y_10:  0.4295
y:  0.5413411329464508
error:  0.11184113294645082


### Solucion:

Utiliza esta tabla como guia para validar que tu funcion este correcta,
y reemplaza los valores que faltan con la salida de tu funcion.

| x     | 0 | 0.1         | 0.2    | 0.4    | 0.6    | 0.8    | 1            |
|-------|---|-------------|--------|--------|--------|--------|--------------|
| y     | 4 | 3.2749      | 2.6813 | 1.7973 | 1.2048 | 0.8076 | 0.5413       |
| $y_n$ | 4 | $y_1$       | 2.56   | 1.6384 | 1.0486 | 0.6711 | $y_10$       |
| error | 0 | y - $y_1$   | 0.1213 | 0.1589 | 0.1562 | 0.1365 | y-$y_{10}$ |

### 4.1 Probando la implementacion con una aplicación real de data science

Podemos utilizar ecuaciones diferenciales  como modelos matemáticos de fenónemos reales, por ejemplo si llegas a trabajar como data scientist para una farmaceutica  puede que interese analizar la desintegración radioactiva de cierta sustancia y como esto cambia en el tiempo
#### Desintegracion Radiactiva

$A$ es la cantidad de sustancia en el tiempo $t$, la derivada $\frac{\partial A}{\partial t}$ es rapidez a la que se desintegran los nucleos de una sustancia.

Queremos aproximar la solucion particular de la ecuacion diferencial
\begin{equation*}
\ \frac{\partial A}{\partial t} =  -0.56A \\
\end{equation*}
Donde $A(0) = 1$ vamos a un tamaño de $h=0.1$

Tambien vamos a calcular el error en este caso podemos verificar el error  con la solucion exacta:
\begin{equation*}
\ A(t) = e^{-0.56t} \\
\end{equation*}

**Nota** En este caso sencillo, poseemos la solución exacta a la ED por lo cual no se necesita una solución aproximada , pero en aplicaciones y proyectos reales muchas veces la solución exacta no se posee, o es muy difícil de calcular y recurrimos a soluciones aproximadas por computadora , para este ejercicio el saber la solución exacta nos sirve para medir la exactitud de nuestra solución aproximada.

## 5 - Rotacion de listas hacia la izquierda

Una rotacion a la izquierda en una lista mueve cada elemento de la lista una vez hacia la izquierda. Por ejemplo, si hacemos $1$ rotacion en en el arreglo $[1,2,3,4,5]$, el arreglo resultante seria $[2,3,4,5,1]$

#### Descipcion de la funcion
Dado un arreglo $a$ de $n$ integers y un numero $d$ de rotaciones, realizar $d$ rotaciones hacia la izquierda. La funcion debe retorner el arreglo re-ordenado.

**rotIzquierda** tiene los siguientes parametros:

* $a$ Un arreglo de numeros entreros.
* $d$ numero de rotaciones.

In [31]:
 def rotIzquierda(a, d):
    return a[d:] + a[:d]

In [32]:
rotIzquierda([1, 2, 3, 4, 5], 4)
# Resultado: [5, 1, 2, 3, 4]

[5, 1, 2, 3, 4]

### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

##### Importante :
En este ejercicio se evalua que la funcion sea optima, el tiempo de ejecucion tiene que ser el menor posible, se va a evaluar el resultado correcto y el tiempo de ejecucion, si en alguna de las pruebas obtienes **[FALLO: TIMEOUT]** la funcion no es optima debes cambiarla hasta que todas las pruebas pasen

#### TIP:
Utiliza **slice** en el array para optimizar la funcion

In [33]:
validaciones.rotIzquierda_function(rotIzquierda)

Prueba 1
Prueba 1[CORRECTA]
Prueba 2
Prueba 2[CORRECTA]
Prueba 3
Prueba 3[CORRECTA]
